Add RuleKind to RuleConfiguration

marcelofabri opened the pull request:
Partially fixes #1282

In some of the rules I was in doubt between 2 kinds, so this can/should be validated.

jpsim commented:
I definitely like this direction, that's the right way to do this IMO./

norio-nomura commented:
> In some of the rules I was in doubt between 2 kinds, so this can/should be validated.

If Kind would be protocol, rule could be adapt to multiple kinds.

marcelofabri commented:
> If Kind would be protocol, rule could adapt to multiple kinds.

My initial feeling is that from an user perspective it's better that a rule can belong to only one kind, even if some of them are in a "gray area".

masters3d commented on Source/SwiftLintFramework/Protocols/Rule.swift:
```diff
@@ -8,10 +8,19 @@
     init() // Rules need to be able to be initialized with default values
     init(configuration: Any) throws
     static var description: RuleDescription { get }
+    static var kind: RuleKind { get }
```
What about using an option set? I think the advantage is that a rule can be classifed with multiple "labels" without those labels having to be protocols. I like the idea of protocols as labels too. 
The emum restriction kind of reminds me of  http://www.jessesquires.com/enums-as-configs/

marcelofabri replied:
I still think that from an user point of view, it's better that a rule can only belong to one category.

This way we could easily in the future allow enabling/disabling full categories that once. That would be tricky if a rule was in two groups: what would be the expected behavior?

masters3d replied:
What about config options that change behavior?

marcelofabri replied:
I'm not sure I understand, can you elaborate?

masters3d replied:
Rules with configurations that make the rule be a different category depending on what config is given.

marcelofabri replied:
Can you think of any examples?

masters3d replied:
Colon, Comma.

marcelofabri replied:
for me they'd be style rules no matter the configuration

marcelofabri commented:
@jpsim Is this something that we want to move forward with? 

Revisiting this, I'd add `kind` in `RuleConfiguration` instead.

marcelofabri commented:
I've rebased this and moved the definition to `RuleDescription`. Basically now we should:

- [ ] Validate that the kinds (do they meaningfully divide the rules?)
- [ ] Validate each rule kind
- [ ] Choose what to do with custom rules. Currently it's always flagged as `.style`. Options:
    * Allow `kind` to be specified in configuration? (What if it's not specified?)
    * Create a new kind `.custom`?
